### DISCUSSION TITLE
Long in GraphQL

### DISCUSSION THREAD

**Discussion by uPagge** [quarkusio/quarkus#24985] [Q&A]
When I try to use the Long type, I get a conversion to BigInteger. But it so happened historically that I need to get exactly Long. AdaptToScalar not working

<img width="582" alt="Снимок экрана 2022-04-18 в 15 48 23" src="https://user-images.githubusercontent.com/40397740/163810583-b3616d26-e9fa-4a2a-8653-2f5242c9027e.png">

<img width="635" alt="Снимок экрана 2022-04-18 в 15 48 41" src="https://user-images.githubusercontent.com/40397740/163810620-d93d2512-a2e4-4700-8724-3fb11b80f6e4.png">

[a comment by an automated account is not shown]

**Reply by phillip-kruger** [reply]
I think the problem is that `@AdaptToScalar` does not work with input parameters. We need to fix that. In the mean time you can try and use an input object that contains the fields and then use `AdaptToScalar` there.

**Reply by uPagge** [reply]
> I think the problem is that `@AdaptToScalar` does not work with input parameters. We need to fix that. In the mean time you can try and use an input object that contains the fields and then use `AdaptToScalar` there.

As far as I understand for input, I will have to change the graphql schema, and I can't influence it.
She looks something like this

```
type Mutation {
     update(id: Long!, request: UpdateUserRequest!): UpdateUserResponse!
}
...
scalar Long
```

**Reply by uPagge** [reply]
I tried using Dynamic client. But I ran into another problem. I need to send a Header with a session ID in the request.

https://quarkus.io/guides/smallrye-graphql-client#using-the-dynamic-client

The documentation suggests using this format

`quarkus.smallrye-graphql-client.star-wars-dynamic.header.HEADER-KEY=HEADER-VALUE"`

But the session is requested and changed while the application is running. Therefore, it is not clear how this can be implemented, except to build a new `VertxDynamicGraphQLClient` each time using `VertxDynamicGraphQLClientBuilder`. But this is not the most convenient solution.

**Reply by jmartisk** [reply]
@uPagge , unfortunately building a new client using the builder is the only currently possible solution. We could potentially add an SPI that allows adding a "header provider" or something

**Reply by uPagge** [reply]
@jmartisk It would be great if this appeared in future versions. Thanks

**Reply by uPagge** [reply]
@jmartisk I have another thought about this. Perhaps I could update the property value at runtime? But a cursory examination of the documentation shows that this is impossible. I found only one way, using

org.microprofile-ext.config-ext:configsource-memory

But the author uses REST to update properties, I would not want to set such an endpoint

Source https://sgitario.github.io/quarkus-update-custom-runtime/

Is there another way to change the property value at runtime?

**Reply by phillip-kruger** [reply]
I am that author :) you can write your own config source?

**Reply by jmartisk** [reply]
Changing a property value is one thing, but the problem is that the client instance is built with a particular value and doesn't look into the configuration properties anymore, so it won't see the update.

We could change it so that the client always looks again into the configuration, but it's not as flexible as having an SPI for header provider. And it's also not really good for performance, I think re-reading the configuration is quite expensive.

**Reply by uPagge** [reply]
Yes, the configuration option is not very successful, and it looks like a crutch, but this is the first thing that came to my mind :)

As a result, so far I have done everything using the Dynamic client.

**Reply by jmartisk** [reply]
I've reported https://github.com/smallrye/smallrye-graphql/issues/1368 for dynamically supplying headers, let's continue the discussion about it in that issue.

**Comment by gsmet** [comment]
@phillip-kruger I don’t know where the first screenshot is from but if it’s in our code, I find it weird we translate Longs to BigInteger. Is there a reason for that?

**Reply by uPagge** [reply]
This screenshot from

io.smallrye.graphql.client.impl.typesafe.reflection.ParameterInfo

dependency:
io.smallrye:smallrye-graphql-client:1.4.3

**Reply by phillip-kruger** [reply]
There is no Long in GraphQL (see https://graphql.org/learn/schema/#scalar-types) by default. Int is too small for Long. We added a few more scalars on the server side (see https://github.com/smallrye/smallrye-graphql/blob/main/server/implementation/src/main/java/io/smallrye/graphql/scalar/number/BigIntegerScalar.java) and Long gets map to BigInteger by default. We did not want to add too much new scalar types. We do not have a way (yet) to easily add your own scalars (it's on our TODO, see  https://github.com/smallrye/smallrye-graphql/issues/911). On the client side, I am not sure if there is a way to hack this. Maybe @jmartisk can comment ?

**Reply by jmartisk** [reply]
@t1 might know better because he wrote the relevant typesafe client code, but I think it's not possible. 
But we certainly could use a way to denote the input types that are expected to be on the server side. A type represented by a class can be annotated with `@Input` to denote the type name, but that annotation is not applicable to a parameter, which is what we would need here to be able to annotate the `Long id` parameter.

**Reply by jmartisk** [reply]
But I'm not very happy with the idea of adding a new annotation which does almost the same as `@Input`, just making it applicable to a parameter... Any better ideas? :/

**Reply by phillip-kruger** [reply]
Use @Input on the parameter ?

**Reply by t1** [reply]
Hmmm... all parameters are Input, aren't they? All return values are Type.

**Reply by jmartisk** [reply]
Parameters are inputs, but declaring the input type name via `@Input` is not possible because that annotation can't be applied to parameters. And it's from the MP spec so we can't simply change it. 

- We could add another annotation, but that would make things messy. 
- Or we could make it a configuration thing, where we allow the user to pass a configuration map that defines the mapping, like: `java.lang.Long` (Java class) -> `Long` (GraphQL type), which is not much nicer.

**Reply by phillip-kruger** [reply]
It should change on the spec level. If need be we can add a input in smallrye API until it can move to the spec?
